sqlmesh/core/context.py

```diff
@@ -1597,53 +1598,200 @@ def table_diff(
-            ).qualified_view_name.for_environment(source_env.naming_info, adapter.dialect)
-
-            target = next(
-                snapshot for snapshot in target_env.snapshots if snapshot.name == model.fqn
-            ).qualified_view_name.for_environment(target_env.naming_info, adapter.dialect)
-
-            source_alias = source_env.name
-            target_alias = target_env.name
-
-            if not on:
-                on = []
-                for expr in [ref.expression for ref in model.all_references if ref.unique]:
-                    if isinstance(expr, exp.Tuple):
-                        on.extend(
-                            [key.this.sql(dialect=adapter.dialect) for key in expr.expressions]
+                raise SQLMeshError(f"Could not find environment '{target}'")
+
+            selected_models = self._new_selector().expand_model_selections(select_models)
+            models_to_diff: t.List[
+                t.Tuple[Model, EngineAdapter, str, str, t.Optional[t.List[str] | exp.Condition]]
+            ] = []
+            models_in_source: t.List[str] = []
+            models_in_target: t.List[str] = []
+            models_no_diff: t.List[str] = []
+            models_without_grain: t.List[Model] = []
+            source_snapshots_to_name = {
+                snapshot.name: snapshot for snapshot in source_env.snapshots
+            }
+            target_snapshots_to_name = {
+                snapshot.name: snapshot for snapshot in target_env.snapshots
+            }
+
+            for model_fqn in selected_models:
+                model = self._models[model_fqn]
+                adapter = self._get_engine_adapter(model.gateway)
+                source_snapshot = source_snapshots_to_name.get(model.fqn)
+                target_snapshot = target_snapshots_to_name.get(model.fqn)
+
+                if source_snapshot is None and target_snapshot:
+                    models_in_source.append(model_fqn)
+                elif target_snapshot is None and source_snapshot:
+                    models_in_target.append(model_fqn)
+                elif target_snapshot and source_snapshot:
+                    if source_snapshot.fingerprint != target_snapshot.fingerprint:
+                        # Compare the virtual layer instead of the physical layer because the virtual layer is guaranteed to point
+                        # to the correct/active snapshot for the model in the specified environment, taking into account things like dev previews
+                        source = source_snapshot.qualified_view_name.for_environment(
+                            source_env.naming_info, adapter.dialect
                         )
+                        target = target_snapshot.qualified_view_name.for_environment(
+                            target_env.naming_info, adapter.dialect
+                        )
+
+                        model_on = []
```

Comment:
This block looks like it should be done in a separate function. Maybe a method in the model definition (or wherever we define `all_references`).